qubes/storage/callback.py

```diff
@@ -0,0 +1,648 @@
+        '''
+        #NOTE: attribute names **must** start with `_cb_` unless they are meant to be stored as self._cb_impl attributes
+        self._cb_ctor_done = False #: Boolean to indicate whether or not `__init__` successfully ran through.
+        self._cb_log = logging.getLogger('qubes.storage.callback') #: Logger instance.
+        if not isinstance(conf_id, str):
+            raise qubes.storage.StoragePoolException('conf_id is no String. VM attack?!')
+        self._cb_conf_id = conf_id #: Configuration ID as passed to `__init__()`.
+
+        config_path = '/etc/qubes_callback.json'
+        with open(config_path) as json_file:
+            conf_all = json.load(json_file)
+        if not isinstance(conf_all, dict):
+            raise qubes.storage.StoragePoolException('The file %s is supposed to define a dict.' % config_path)
+
+        try:
+            self._cb_conf = conf_all[self._cb_conf_id] #: Dictionary holding all configuration for the given _cb_conf_id.
+        except KeyError:
+            #we cannot throw KeyErrors as we'll otherwise generate incorrect error messages @qubes.app._get_pool()
+            raise qubes.storage.StoragePoolException('The specified conf_id %s could not be found inside %s.' % (self._cb_conf_id, config_path))
+
+        try:
+            bdriver = self._cb_conf['bdriver']
+        except KeyError:
+            raise qubes.storage.StoragePoolException('Missing bdriver for the conf_id %s inside %s.' % (self._cb_conf_id, config_path))
+
+        self._cb_cmd_arg = json.dumps(self._cb_conf, sort_keys=True, indent=2) #: Full configuration as string in the format required by _callback().
+
+        try:
+            cls = qubes.utils.get_entry_point_one(qubes.storage.STORAGE_ENTRY_POINT, bdriver)
+        except KeyError:
+            raise qubes.storage.StoragePoolException('The driver %s was not found on your system.' % bdriver)
+
+        if not issubclass(cls, qubes.storage.Pool):
+            raise qubes.storage.StoragePoolException('The class %s must be a subclass of qubes.storage.Pool.' % cls)
+
+        self._cb_requires_init = self._check_init() #: Boolean indicating whether late storage initialization yet has to be done or not.
+        self._cb_init_lock = asyncio.Lock() #: Lock ensuring that late storage initialization is only run exactly once.
+        self._cb_sync_lock = asyncio.Lock() #: Lock to prevent sync code from running async.
+        bdriver_args = self._cb_conf.get('bdriver_args', {})
+        self._cb_impl = cls(name=name, **bdriver_args) #: Instance of the backend pool driver.
+
+        super().__init__(name=name, revisions_to_keep=int(bdriver_args.get('revisions_to_keep', 1)))
+        self._cb_ctor_done = True
+        self._callback_nocoro('post_ctor')
+
+    @asyncio.coroutine
+    def _coro_maybe(self, value):
+        if asyncio.iscoroutine(value):
+            return (yield from value)
+        with (yield from self._cb_sync_lock):
+            return value
```

Comment:
This doesn't run that function under the lock, it simply takes a lock and instantly releases it (as `return value` doesn't call anything).
I'm not sure what exactly you're trying to protect here  (two synchronous functions can't in python run in parallel, by definition - synchronous function blocks python interpreter until it finishes), but if you want to ensure no sync function is running in parallel with some async one of the same driver (in some unusual case where the driver implements some sync and some async), then you should obtain the `value` (like the call to `self._cb_impl.setup()`) under the lock in the first place.

Reply:
I wanted to fix this issue:
https://github.com/QubesOS/qubes-core-admin/pull/354/commits/287a4a0429165f9fa969960dafbd810a29c421d9#diff-271f0ec3445d2fb476cb711090e750b8L50-L53

I.e. I wanted to make my async code block qubesd as the sync code would do it / imitate the behaviour in order not to break any assumptions made by sync code authors.

As you pointed out an `asyncio` lock cannot achieve that though (i.e. that commit was useless regardless of the other mistakes).

I'm not sure whether Qubes OS runs async code in another process or whether sync & async code share the same.

In the former case (one common process for sync & async) this could only be fixed by blocking the Qubes OS sync main loop entirely during execution / run exclusively. Is there such a "global lock"?

In the latter case (one process each) there's probably no issue at all: Sync code cannot run in parallel to async code and sync code will not be interrupted even if run inside a coroutine as there's no awaitable afaik, i.e. the "word of caution" could be removed.

Please let me know what you think or whether all of this was just plain wrong...